src/common/resourcemonitor/resourcemonitor.cpp

```diff
@@ -0,0 +1,219 @@
+            mNotifySendMonitoring = true;
+            mMonitoringCondVar.NotifyAll();
+        },
+        false);
+}
+
+Error ResourceMonitor::StopSendMonitoring()
+{
+    LOG_DBG() << "Stop send monitoring";
+
+    return mTimerSenders.Stop();
+}
+
+Error ResourceMonitor::GetNodeInfo(NodeInfo& nodeInfo) const
+{
+    LOG_DBG() << "Get node info";
+
+    return AOS_ERROR_WRAP(mResourceUsageProvider->GetNodeInfo(nodeInfo));
+}
+
+Error ResourceMonitor::StartInstanceMonitoring(const String& instanceID, const InstanceMonitorParams& monitoringConfig)
+{
+    LockGuard lock(mMutex);
+
+    LOG_DBG() << "Start instance monitoring";
+
+    auto findInstance = mNodeMonitoringData.mServiceInstances.Find(
+        [&instanceID](const InstanceMonitoringData& instance) { return instance.mInstanceID == instanceID; });
+
+    if (!findInstance.mError.IsNone()) {
+        MonitoringData monitoringData;
+
+        for (const auto& disk : monitoringConfig.mPartitions) {
+            monitoringData.mDisk.EmplaceBack(disk);
+        }
+
+        mNodeMonitoringData.mServiceInstances.EmplaceBack(instanceID, monitoringConfig.mInstanceIdent, monitoringData);
+
+        return ErrorEnum::eNone;
+    }
+
+    findInstance.mValue->mMonitoringData.mDisk.Clear();
+
+    for (const auto& disk : monitoringConfig.mPartitions) {
+        findInstance.mValue->mMonitoringData.mDisk.EmplaceBack(disk);
+    }
+
+    return ErrorEnum::eNone;
+}
+
+Error ResourceMonitor::StopInstanceMonitoring(const String& instanceID)
+{
+    LockGuard lock(mMutex);
+
+    LOG_DBG() << "Stop instance monitoring";
+
+    return mNodeMonitoringData.mServiceInstances
+        .Remove([&instanceID](const InstanceMonitoringData& instance) { return instance.mInstanceID == instanceID; })
+        .mError;
+}
+
+ResourceMonitor::~ResourceMonitor()
+{
+    {
+        LockGuard lock(mMutex);
+        mFinishMonitoring = true;
+        mMonitoringCondVar.NotifyAll();
+    }
+
+    mThreadMonitoring.Join();
+    mThreadSendMonitoring.Join();
+
+    auto err = mTimerSenders.Stop();
+    if (!err.IsNone()) {
+        LOG_ERR() << "Failed to stop sender timer: " << err;
+    }
+
+    err = mTimerGatherers.Stop();
+    if (!err.IsNone()) {
+        LOG_ERR() << "Failed to stop gatherer timer: " << err;
+    }
+}
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+Error ResourceMonitor::RunSendMonitoringData()
+{
+    return mThreadSendMonitoring.Run([this](void*) {
+        while (true) {
+
+            aos::LockGuard lock(mMutex);
+
+            if (lock.GetError() != ErrorEnum::eNone) {
+                LOG_ERR() << "Failed to lock mutex: " << lock.GetError();
+
+                continue;
+            }
+
+            mMonitoringCondVar.Wait([this] { return mFinishMonitoring || mNotifySendMonitoring; });
```

Comment:
In this scenario, I don't see advantages of using cond vars compared to simple sleep. Sleep is better in term of resource usage.

Reply:
Removed. Use sleep instead of timers